### PR TITLE
WIP: Multiple font size / line height spans (updated)

### LINKED ISSUE #64: Spans of font sizes
This is a highly requested feature from a rich text editor - being able to change font size within a line.

I looked into doing this myself but couldn't figure out a good way to get the attrs into layout without adding a font size field to the shaping structs, which seemed to be against the spirit of the design. If there's any advice on what changes would need to be made where to make this happen, I'll gladly do it.

### PR DESCRIPTION
**Pull request by nicoburns**
Implement specifying multiple font sizes and/or line heights.

- Continued from https://github.com/pop-os/cosmic-text/pull/150.
- Fixes #64.

### Current status

Almost done. Working, with some fixes for the editing required.

### Todo

* [x]  fix scrolling
* [x]  implement `enum LineHeight`
* [x]  cache `line_heights()`
* [x]  retire `Metrics`
* [x]  ~rebase~ `main` merged into branch. Will need to squash prior to merge.
* [x]  migrate tests
* [x]  migrate benches
* [x]  migrate example terminal (NOTE: does not compile on Windows with `termion`, see [Update terminal example using olored` #129](https://github.com/pop-os/cosmic-text/pull/129))
* [ ]  migrate example editor-orbclient
  * [ ]  fix bug empty buffer does not render cursor
  * [ ]  fix bug empty buffer has default Attrs
  * [ ]  support rescaling font sizes on demand, perhaps via mutable access to spans of Attrs
* [ ]  migrate example editor-test
* [ ]  migrate example editor-libcosmic (NOTE: does not compile on Windows with pinned version of `iced` (incompatible API))
* [ ]  migrate vi feature
* [ ]  fix editor action Vertical
* [ ]  fix some editor actions not redrawing
* [x]  tidy up
* [ ]  docs and PR write-up

Future work:

* Split `Attrs`: `color_opt`, `metadata`, `font_size`, `line_height`
* `Buffer` should have a default `Attrs`
* Convenience method for updating all spans of `Attrs` in a `Buffer`
* Handle start-of-line `Attrs` more carefully

Questions:

* Should the `Hash` impl of `Attrs` include line height?
* Should `Metrics` be restored as a property of `Attrs`, with definition `Metrics { font_size: f32, line_height: LineHeight }`?

### CODE REVIEW

**Comment by TotalKrill** [comment]
whats missing in this? :)

**Comment by nicoburns** [comment]
> whats missing in this? :)

The checkbox list in the description is accurate. The main thing is the editor abstraction isn't quite working properly (compare for example pressing enter a bunch of times in the `editor-libcosmic` example on `main` vs. this PR - on main you get many blank new lines and the cursor always renders, on this PR you only get one newline and the cursor disappears).

If you have time to look into this your help would be very welcome. I've done some basic fixing up of compile errors. But haven't had time to look into the bugs.

**Comment by TotalKrill** [comment]
Did a quick check, the bug with no cursor on newllne, is because the height of empy lines becomes zero from the LayoutLine::line_height() return the default value of f32 when the line contains no characters... 

I was planning on changing it into returning an option, so empty lines return a None value, but I am unsure of where to grab the line height value from then... Probably the previous value somehow

**Comment by TotalKrill** [comment]
And there is also an error in the [Buffer:set_rich_text] code, its trying to add only newlines to the rich_text example makes that one bug out as well

**Comment by TotalKrill** [comment]
I dont like the segregation, so an update to this PR is found [here](https://github.com/nicoburns/cosmic-text/pull/1)

File: src/attrs.rs [hunk 1]
```diff
@@ -132,6 +180,31 @@ impl<'a> Attrs<'a> {
         self
     }
 
+    /// Set font size
+    ///
+    /// # Panics
+    ///
+    /// Will panic if font size is zero.
+    pub fn size(mut self, size: f32) -> Self {
```
Comment (StaffEngineer) [inline src/attrs.rs:188]: nitpick: `font_size`

Comment (StaffEngineer) [inline src/attrs.rs:188]:
```suggestion
    pub fn font_size(mut self, size: f32) -> Self {
```

**Comment by StaffEngineer** [comment]
I've tested `nicoburns:multi-size-fix-examples` and `TotalKrill:multi-size-fix-examples` using my bevy plugin: https://github.com/StaffEngineer/bevy_cosmic_edit/tree/new-cosmic

Seems like there is a bug after rebasing `tigregalis:multi-size`... basically calling `set_text` with `"Blah \n blah"` creates buffer with 2 buffer lines as expected but `attrs_list.spans` calculated incorrectly for those lines.

**Comment by TotalKrill** [comment]
I am currently testing this using the ./rich-text.sh in this repo. Could you elaborate on the issue you are seeing and would i be reproducable in this repository?

**Comment by StaffEngineer** [comment]
@TotalKrill there is actually the issue on your fork. I created reproducible example here: https://github.com/TotalKrill/cosmic-text/compare/multi-size-fix-examples...StaffEngineer:cosmic-text:multi-size-fix-examples

If you run rich-text example it prints:
```
[examples/rich-text/src/main.rs:58] line.text() = "Blah"
[examples/rich-text/src/main.rs:59] line.attrs_list() = AttrsList {
    defaults: AttrsOwned {
        color_opt: None,
        family_owned: Name(
            "Times New Roman",
        ),
        stretch: Normal,
        style: Normal,
        weight: Weight(
            400,
        ),
        metadata: 0,
        font_size: 54.34,
        line_height: Absolute(
            74.7175,
        ),
    },
    spans: {},
}
[examples/rich-text/src/main.rs:58] line.text() = "blah"
[examples/rich-text/src/main.rs:59] line.attrs_list() = AttrsList {
    defaults: AttrsOwned {
        color_opt: None,
        family_owned: Name(
            "Times New Roman",
        ),
        stretch: Normal,
        style: Normal,
        weight: Weight(
            400,
        ),
        metadata: 0,
        font_size: 54.34,
        line_height: Absolute(
            74.7175,
        ),
    },
    spans: {
        0..9: AttrsOwned {
            color_opt: None,
            family_owned: Name(
                "Times New Roman",
            ),
            stretch: Normal,
            style: Normal,
            weight: Weight(
                400,
            ),
            metadata: 0,
            font_size: 54.34,
            line_height: Absolute(
                74.7175,
            ),
        },
    },
}
```

But it should output:
```
[examples/rich-text/src/main.rs:58] line.text() = "Blah"
[examples/rich-text/src/main.rs:59] line.attrs_list() = AttrsList {
    defaults: AttrsOwned {
        color_opt: None,
        family_owned: Name(
            "Times New Roman",
        ),
        stretch: Normal,
        style: Normal,
        weight: Weight(
            400,
        ),
        metadata: 0,
        font_size: 54.34,
        line_height: Absolute(
            74.7175,
        ),
    },
    spans: {},
}
[examples/rich-text/src/main.rs:58] line.text() = "blah"
[examples/rich-text/src/main.rs:59] line.attrs_list() = AttrsList {
    defaults: AttrsOwned {
        color_opt: None,
        family_owned: Name(
            "Times New Roman",
        ),
        stretch: Normal,
        style: Normal,
        weight: Weight(
            400,
        ),
        metadata: 0,
        font_size: 54.34,
        line_height: Absolute(
            74.7175,
        ),
    },
    spans: {},
}
```
(notice spans are empty)

**Comment by TotalKrill** [comment]
Updated with a fix to that, I believe :)

**Comment by nicoburns** [comment]
Closing in favour of #235 (which is this PR rebased on top of latest `main`)

**Comment by nicoburns** [comment]
Closing in favour of #235 (which is this PR rebased on top of latest `main`)

**Pull request closed without merging**